### PR TITLE
[WMS][12.0] Add stock_move_location_dest_constraint_tag - alpha version

### PR DESCRIPTION
**Pull request by grindtildeath**
This module implements destination location constraint on stock move line.

When installed stock moves will only be allowed if the destination location
includes the same tags as defined on the product or its product category.

N.B.: If tags are defined on the product, tags defined on its product category
will not be considered

Related to #691

Requires:
 - [ ] #696

### CODE REVIEW

File: stock_move_location_dest_constraint_tag/models/stock_location.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    _inherit = 'stock.location'
+
+    product_storage_tag_ids = fields.Many2many(
+        'product.storage.tag',
+        string='Storage Tags',
+    )
+
+    def check_move_dest_constraint(self, line=None, product=None):
+        res = super().check_move_dest_constraint(product)
```
Comment (grindtildeath) [inline stock_move_location_dest_constraint_tag/models/stock_location.py]:
```suggestion
        res = super().check_move_dest_constraint(line=line, product=product)
```

File: stock_move_location_dest_constraint_tag/demo/product.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
```
Comment (Trivedi-Vacha-SerpentCS) [inline stock_move_location_dest_constraint_tag/demo/product.xml:2]: should use odoo noupdate="1"

File: stock_move_location_dest_constraint_tag/demo/product_category.xml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
```
Comment (Trivedi-Vacha-SerpentCS) [inline stock_move_location_dest_constraint_tag/demo/product_category.xml:2]: noupdate="1" is missing for all demo files

File: stock_move_location_dest_constraint_tag/models/product_category.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import models, fields
```
Comment (Trivedi-Vacha-SerpentCS) [inline stock_move_location_dest_constraint_tag/models/product_category.py:3]: import should be in alphabetical order
 from odoo import fields, models

File: stock_move_location_dest_constraint_tag/models/product_storage_tag.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import models, fields
```
Comment (Trivedi-Vacha-SerpentCS) [inline stock_move_location_dest_constraint_tag/models/product_storage_tag.py:3]: same as above

File: stock_move_location_dest_constraint_tag/models/product_template.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import models, fields
```
Comment (Trivedi-Vacha-SerpentCS) [inline stock_move_location_dest_constraint_tag/models/product_template.py:3]: same as above

File: stock_move_location_dest_constraint_tag/models/stock_location.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import models, fields, _
```
Comment (Trivedi-Vacha-SerpentCS) [inline stock_move_location_dest_constraint_tag/models/stock_location.py:3]: It should be from odoo import _, fields, models

**Review by gurneyalex — COMMENTED** [review]
I'm concerned about a few issues here:

* with this implementation, the Customers location needs to be updated with every possible tag, otherwise you will never be able to deliver a SO
* same for PACK and OUT locations of a WH
* there is no domain update in the choice of a destination location to help the user choose an appropriate location.

**Comment by jgrandguillaume** [comment]
@gurneyalex I agree with you. We need to review all this put away design at a whole. The POC prooved it works, but it's not optimal.

**Pull request closed without merging**